### DISCUSSION TITLE
Download functionality

### DISCUSSION THREAD

**Discussion by psinger** [h2oai/wave#838] [Q&A]
I have the following use case:

- A (large) file exists on disk
- I want to enable the user of the app to download it
- I upload the file to the web server via `q.site.upload`
- I add a download button via `ui.link` to enable to download it

However, as the file can be quite large, I always have to run upload first, even though the file exists on disk. This operation can be quite slow. I now have two questions:

1. Is it possible to enable the user to download directly from disk without running `q.site.upload` first? I suppose not.
2. If not, can I only trigger the upload as soon as the user clicks the download button? Otherwise I always have to run the slow upload first upfront, without me knowing if the user will even download it. I could write a separate routine, but I do not know how to trigger a download without `ui.link`.

**Comment by lo5** [comment]
One possible solution to this is:
- Allow something like `remote_path = q.site.uplink(local_path)`, which makes the server aware of the file, but does not actually upload it.
- Make the server pull the file from the app only when someone requests it from the browser.

This solution will not require you to do anything extra in your app other than use `q.site.uplink()` instead of `q.site.upload()`. The framework can take care of serving the file when requested until you call `q.site.unlink()` to stop serving the file. 

Will this address your requirements?

**Reply by psinger** [reply]
@lo5 

Yes, this would address my requirements. But this is not implemented yet, right? How could we get this going?
